core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -54,7 +54,8 @@ class ReplicaFetcherThread(name: String,
                                 sourceBroker = sourceBroker,
                                 fetchBackOffMs = brokerConfig.replicaFetchBackoffMs,
                                 isInterruptible = false,
-                                includeLogTruncation = true) {
+                                includeLogTruncation = true,
+                                useLeaderEpochInResponse = brokerConfig.interBrokerProtocolVersion >= KAFKA_2_0_IV0) {
```

Comment:
In ReplicaFetcherThread.fetchEpochsFromLeader(), the version of OffsetsForLeaderEpochRequest should probably be determined based on the interBrokerProtocolVersion. We can use OffsetsForLeaderEpochRequest V1 only if the interBrokerProtocolVersion >= KAFKA_2_0_IV0. Otherwise, we can rolling bounce the cluster to upgrade the code, the leader may still be running the old code and not recognizes OffsetsForLeaderEpochRequest V1.

Reply:
The OffsetsForLeaderEpoch request is exactly the same in v0, so we don't need to explicitly check the protocol version when building the requests. If the leader is on older version, it will send v0 response, which will not include leader epoch field, which is handled in OffsetsForLeaderEpochResponse constructor by setting leader epoch field to undefined. In the fetcher thread, we handle this case (where the leader epoch is undefined) in maybeTruncate() and fall back to KIP-101 behavior, same as when this broker is on older protocol version.

Reply:
Hmm.. my understanding is that the version of the response should always match the version of the request. Thus in order to receive OffsetsForLeaderEpochResponse V1, the broker needs to send OffsetsForLeaderEpochRequest V1. And the broker should request the request if the version of the request is not recognized. Did I miss something?

Reply:
Yes, correct. I meant nothing different to do in the fetcher thread. If I understood the code correctly, ReplicaFetcherThread.fetchEpochsFromLeader() passes the OffsetsForLeaderEpochRequest.Builder to sendRequest(), and then build() is called on that builder with with a version in NetworkClient.doSend. It looks like the proper version will be used in that case.

Reply:
Currently if we do not explicitly specify the version for AbstractRequest.Builder(), the latest version of this request, as determined by ApiKeys.latestVersion(), will be used for this request. The latest version for OffsetsForLeaderEpochRequest will V1 after this patch. we probably need to explicitly pass the version (determined by the IBP) to OffsetsForLeaderEpochRequest.Builder, similar to what we do for UpdateMetadataRequest.Builder() in ControllerBrokerRequestBatch.sendRequestsToBrokers().

Reply:
Oh I see, thank you, let me take a look.

Reply:
Thanks a lot for your help, I updated the code to use the OffsetsForLeaderEpochRequest version when building a request.